EIPS/eip-7496.md

```diff
@@ -0,0 +1,216 @@
+
+If `valueMappings` is not used, the default trait values for `boolean` should be `bytes32(0)` for `false` and `bytes32(uint256(1))` (`0x0000000000000000000000000000000000000000000000000000000000000001`) for `true`.
+
+#### `epochSeconds` Metadata Type
+
+The `epochSeconds` metadata type allows for a numeric value to be set for a trait in seconds since the Unix epoch.
+
+The `dataType` object MAY have a `valueMappings` object defined. If the `valueMappings` object is defined, the `valueMappings` object MUST be a mapping of `bytes32` values to integer or unset `null` values.
+
+### Events
+
+Updating traits MUST emit one of:
+
+- `TraitUpdated`
+- `TraitUpdatedRange`
+- `TraitUpdatedRangeUniformValue`
+- `TraitUpdatedList`
+- `TraitUpdatedListUniformValue`
+
+For the `Range` events, the `fromTokenId` and `toTokenId` MUST be a consecutive range of tokens IDs and MUST be treated as an inclusive range.
+
+For the `List` events, the `tokenIds` MAY be in any order.
+
+It is RECOMMENDED to use the `UniformValue` events when the trait value is uniform across all token ids, so offchain indexers can more quickly process bulk updates rather than fetching each trait value individually.
+
+Updating the trait metadata URI MUST emit the event `TraitMetadataURIUpdated` so offchain indexers can be notified to query the contract for the latest changes via `getTraitMetadataURI()`.
+
+### `setTrait`
+
+If a trait defines `tokenOwnerCanUpdateValue` as `true`, then the trait value MUST be updatable by the token owner by calling `setTrait`.
+
+If the value the token owner is attempting to set is not valid, the transaction MUST revert. If the value is valid, the trait value MUST be updated and one of the `TraitUpdated` events MUST be emitted.
+
+If the trait has a `valueMappings` entry defined for the desired value being set, `setTrait` MUST be called with the corresponding `traitValue`.
+
+## Rationale
+
+Onchain traits can be used by contracts to get and mutate traits in a variety of different scenarios. For example, a contract that wants to entitle a token to a consumable benefit (e.g. a redeemable) can robustly reflect that onchain. Marketplaces can allow bidding on these tokens based on the trait value without having to rely on offchain state and exposing users to frontrunning attacks.
+
+## Backwards Compatibility
+
+As a new EIP, no backwards compatibility issues are present, except for the point in the specification above that it is explicitly required that the onchain traits MUST override any conflicting values specified by the ERC-721 or ERC-1155 metadata URIs.
+
+## Test Cases
+
+Authors have included Foundry tests covering functionality of the specification in the assets folder.
+
+## Reference Implementation
+
+Authors have included reference implementations of the specification in the assets folder.
```

Comment:
You can link to them if you like. Something like:

```markdown
[assets folder](../assets/eip-7496/DynamicTraits.sol)
```

Reply:
thanks, linked